lib/stages/visRawReader.cpp

```diff
@@ -303,64 +322,52 @@ void visRawReader::main_thread() {
             std::memcpy(frame, mapped_file + file_ind * file_frame_size + metadata_size + 1,
                         data_size);
 
-            // Update the dataset ID. If not using comet, only need to do this once.
-            dset_id = ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->dataset_id;
-            if (!got_first_frame) {
-                get_dataset_state(dset_id);
-                cur_dset_id = dset_id;
-                got_first_frame = true;
-            } else if (dset_id != cur_dset_id) {
-                if (!use_comet) {
-                    WARN("Dataset ID of incoming frames changed from {} to {}. Changing  ID "
-                         "not supported without dataset broker. Ignoring.",
-                         cur_dset_id, dset_id);
-                } else {
-                    get_dataset_state(dset_id);
-                }
-                cur_dset_id = dset_id;
-            }
         } else {
-            // Set metadata if file contained an empty frame
-            ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->num_prod = _prods.size();
-            ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->num_ev = _ev.size();
-            ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->num_elements = _inputs.size();
-            // Fill data with zeros
+            // Create frame and set structural metadata
             size_t num_vis = _stack.size() > 0 ? _stack.size() : _prods.size();
             auto frame = visFrameView(out_buf, frame_id, _inputs.size(), num_vis, _ev.size());
+
+            // Fill data with zeros
             std::fill(frame.vis.begin(), frame.vis.end(), 0.0);
             std::fill(frame.weight.begin(), frame.weight.end(), 0.0);
             std::fill(frame.eval.begin(), frame.eval.end(), 0.0);
             std::fill(frame.evec.begin(), frame.evec.end(), 0.0);
             std::fill(frame.gain.begin(), frame.gain.end(), 0.0);
             std::fill(frame.flags.begin(), frame.flags.end(), 0.0);
-            frame.freq_id = 0;
             frame.erms = 0;
-            frame.dataset_id = out_dset_id;
+
+            // Set non-structural metadata
+            frame.freq_id = 0;
+            frame.dataset_id = dset_id_t::null;
+            frame.time = std::make_tuple(0, timespec{0, 0});
+
             // mark frame as empty by ensuring this is 0
             frame.fpga_seq_length = 0;
+            frame.fpga_seq_total = 0;
+
             DEBUG("visRawReader: Reading empty frame: {:d}", frame_id);
         }
 
         // Set the dataset ID to the updated value
-        ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->dataset_id = out_dset_id;
+        dset_id_t& ds_id = ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->dataset_id;
+        ds_id = get_dataset_state(ds_id);
```

Comment:
I guess this means it will looks through the keys of the `ds_in_file` map for every frame. This is probably a small overhead, but I don't have a clear idea of how many different dataset IDs there can be in one file.

Reply:
Yeah. I figured it was worth the very small amount of overhead to clarify the handling. Lookups in a `std::map` are `O(log N)` so pretty cheap, and at most we might have 10s of datasets per file, and 100s if we go to very fast cadence updates.